packages/discovery-provider/src/api/v1/transactions.py

```diff
@@ -72,23 +79,24 @@ class GetTransactionHistory(Resource):
         params={"id": "A User ID"},
     )
     @full_user_ns.expect(transaction_history_parser)
-    @auth_middleware(transaction_history_parser)
+    @auth_middleware(
+        parser=transaction_history_parser,
+        id_param="id",
```

Comment:
I don't _love_ this `id_param` strategy... It's probably required in order to fit in our current world though, and I think we can still go forward with it - especially since I think we've normed on having the query param by default for most everything else. The concern I have is this seems to couple the decorator with the implementation of the API method and feels a bit too "magical." It also assumes that every API will always specify the user ID as a param somehow if it wants authentication, which I think is valid but also blocks us from ever doing `/me` endpoints etc.

FWIW, I'm thinking of "auth" as two steps/responsibilities:

1. The `auth_middleware` _authenticates_ the user, ensuring they are an Audius user and telling the decorated method with which one
2. The API method checks the _authorization_ of that user, ensures they have access to what they're accessing

You don't have to do this - but my brain was spinning so felt I should share my thoughts:

Thinking back to our discussions on headers vs query params vs route params... I think seeing this now and seeing as our signatures are in headers, maybe it _would_ have made more sense to have the user ID you're impersonating be part of the header too - could even just be part of the message of the signature...

In API method:
```python
# optional authentication
@ns.authenticate(allow_anonymous=True)
def my_method(id, authed_user_id):
    # ...

# required authentication, authorization
@ns.authenticate
def my_method(id, authed_user_id):
    abort_if_forbidden(ns, authed_user_id, id)
    # ...
```

In `@auth_middleware`:
1. Recover signer wallet
2. Find corresponding user
3. Check corresponding user is equal to _or_ can manage the user ID specified in the signed message (or another header)
4. Pass the user ID specified in the signed message (or another header) to the method

This has the downside of not allowing granular management of an account (in this flow, managers effectively impersonate the user they manage) and requiring managers to always specify who they're acting on behalf of, but the upside of minimal amount of changes to the existing API, less "magic", and keeps our routes and query params uncoupled with the auth middleware

DEF not blocking on this, though, I can live with `id_param`

Reply:
These are all great points and the main reason I opened it as a draft for comment before finishing it.
I also don't love the `id_param` field. And I toyed with the idea of just encapsulating the logic we use everywhere into a single function we could sprinkle throughout the code.
I'm trying to balance a handful of things:

1. Relying on the wrapped method to implement the correct logic each time is error-prone and I wanted to move that out to a single place that can just be specified through an access level configuration
2. Many of our endpoints allow you to request information on an id that does not match the logged in user. That id is typically (always?) passed as a URL parameter and there's no great way for the decorator to correctly intuit the name of that parameter unless it's explicitly told
3. We will have to continue to support legacy calls. So even if we started accepting "effective user id" as a header, we still have to support reading it off the URL.

I'm actually more in favor of _not_ doing the permissions check in auth_middleware if we think we can be diligent about calling the utility function when its needed. And I think granular permissions are doable in either strategy. It just requires that we have defined a permission structure and then know the authenticated user and the "contextual" user.


Regarding `/me` endpoints: The legacy endpoints I updated as part of this PR follow that strategy and they don't work well with this decorator. It does pretty much rely on the id for the user being passed either as a route or query param that can be grabbed by the decorator. So it is more limiting that way.


I think what I'm taking away from this is that the decorator/middleware should probably _only_ operate on things it can read directly off the request without help (i.e. auth headers and conventions shared by all APIs). If we wanted to allow the decorator to do both Authn and Authz, we need to pass the effective user id in a header as well.

Reply:
RE: (3), we don't have manager mode released yet, so we could default to using the recovered signer user_id if not present, no? I don't think we would need to read it off the URL in the middleware, but yeah just at the endpoint level